matrix-org/synapse
# Support OIDC backchannel logouts

Fixes #11326 

Based on #11482.

Tested (and works) with Keycloak.

### How to test this?

 - Spin up a Keycloak instance, create a realm
 - Configure Synapse to login via Keycloak: https://matrix-org.github.io/synapse/latest/openid.html#keycloak
 - In Keycloak, on the client used by Synapse, add the following backchannel logout URL: `{public_url}/_synapse/client/oidc/backchannel_logout`
 - Ensure `Backchannel Logout Session Required` is enabled on that client
 - Login via OIDC in any Matrix client
 - Log the user out of Keycloak by going to `{keycloak}/auth/realms/{realms}/account/#/` (which is the self-service console) and click "Sign Out"
 - The user should be logged out of Synapse

There is a Docker Compose stack here to help with testing: https://github.com/sandhose/synapse-keycloak-devstack

Additionally, there are two edge cases which can be tested (and which are covered by tests):

 - when the logout happens before the client sends the login token (e.g. happens during the "Continuing will grant app.element.io access to your account." screen)
 - when the logout happens during the username mapping session, so when the `localpart_template` is not set and the user chooses a username themselves

### Pull Request Checklist



* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))
